add middleware doc for server (#116)
hyperium/hyperium.github.io: _stable/server/middleware.md

---
title: Getting Started with a Server Middleware
layout: guide
---

> As [Upgrade](upgrading) mentioned, hyper v1 is not depended on tower for the Service trait. When we want to add tower-like middleware, there are 2 kind of approach to make it.

Reviewer: ```suggestion
As [Upgrade](upgrading) mentioned, hyper v1 does not depend on tower for the `Service` trait. When we want to add tower-like middleware, there are 2 kinds of approach to make it.
```

    fn call(&self, req: Req) -> Self::Future {
        println!("processing request: {} {}", req.method(), req.uri().path());
        self.inner.call(req)
    }
}
# fn main() {}
```

Then this can be used in server:

```rust
# extern crate tower;
# extern crate hyper;
# extern crate http_body_util;
# extern crate tokio;
# extern crate hyper_util;
# mod no_run {
use std::{convert::Infallible, net::SocketAddr};
use hyper::{
    service::Service,
    body::{Bytes, Incoming},
    server::conn::http1,
    Request, Response,
};
use http_body_util::Full;
use hyper_util::rt::TokioIo;
use tokio::net::TcpListener;
use tower::ServiceBuilder;

# #[derive(Debug, Clone)]
# pub struct Logger<S> {
#     inner: S,
# }
# impl<S> Logger<S> {
#    pub fn new(inner: S) -> Self {
#        Logger { inner }
#    }
# }
# type Req = Request<Incoming>;

# impl<S> Service<Req> for Logger<S>
# where
#     S: Service<Req>,
# {
#     type Response = S::Response;
#     type Error = S::Error;
#     type Future = S::Future;
#     fn call(&self, req: Req) -> Self::Future {
#         println!("processing request: {} {}", req.method(), req.uri().path());
#         self.inner.call(req)
#     }

> # }

Reviewer: This part seems redundant. Maybe just refer `Logger` to the code we've written above for brevity?

Author: Hi @dswij this redundant code block is for passing the [ci doc test](https://github.com/hyperium/hyperium.github.io/blob/master/.github/workflows/check_guides.sh)

 Not sure if is there a way to refer it without failing the doc test.

Reviewer: Not a hard requirement, but I wonder if you can "hide" code lines like in mdbook: https://rust-lang.github.io/mdBook/format/mdbook.html#hiding-code-lines

Author: Hi @dswij, just saw your comment, which part of the code do you think should be hidden in the doc?

Reviewer: This part on the `Logger` makes the example code wordy, IMO. We can probably refer to a full example somewhere at the end, but hiding it here instead of repeating it is easier to read.

Reviewer 2: It looks like the Logger is hidden, each line starts with a `#`.
